### PR TITLE
Confirm payments in embedded

### PR DESCRIPTION
**Pull request by porter-stripe**
## Summary
- Supports confirming payments in the confirm and continue formSheetAction for embedded
- Fixes a bug where selecting a form PM and filling it out, then selecting back to a saved card did not update the payment option correctly. Added a test for this.
- A PR immediately following this will add more UI tests for more edge cases like CVC recollection (which was tested manually)

## Motivation
- Embedded

## Testing
- Manual
- Unit tests
- UI tests

## Changelog
N/A

### CODE REVIEW

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift [hunk 1]
```diff
@@ -18,15 +18,13 @@ import UIKit
...
-    ///   - paymentOption: The `PaymentOption` to be confirmed.
     ///   - completion: A completion handler to call with the `PaymentSheetResult` from the confirmation attempt.
     func embeddedFormViewControllerShouldConfirm(
         _ embeddedFormViewController: EmbeddedFormViewController,
-        with paymentOption: PaymentOption,
         completion: @escaping (PaymentSheetResult, STPAnalyticsClient.DeferredIntentConfirmationType?) -> Void
     )
 
-    /// This method is called when the user taps the primary button (e.g., "Buy") while `formSheetAction` is set to `.continue`.
+    /// This method is called when the user taps the primary button (e.g., "Buy") while `formSheetAction` is set to `.confirm` after confirmation.
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift:27]: 1. Maybe rewrite to:
When `formSheetAction` is set to `.confirm`, this method is called after a user taps the primary button (e.g, "Buy") followed by a successful confirmation.

2. The name of the function is awkward to me.  "xxxxShouldContinue" seems like the function would return a boolean to describe xxxx to continue or not.

3. The name "Continue" clashes with the formSheetAction value .continue -- maybe we can find a better name here?

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift [hunk 2]
```diff
@@ -293,6 +300,68 @@ extension EmbeddedPaymentElement: EmbeddedFormViewControllerDelegate {
...
+                    // formViewController is never currently presented during confirmation in continue mode
+                    if let presentingViewController {
+                        return STPAuthenticationContextWrapper(presentingViewController: presentingViewController)
+                    }
+                    return nil
+                }
+            }()
+            
+            guard let authContext else {
+                return (.failed(error: PaymentSheetError.unknown(debugDescription: "Unexpectedly found nil on both bottomSheetController and presentingViewController.")), STPAnalyticsClient.DeferredIntentConfirmationType.none)
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift:327]: Probably better just to put that authContext is nil.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift [hunk 3]
```diff
@@ -293,6 +300,68 @@ extension EmbeddedPaymentElement: EmbeddedFormViewControllerDelegate {
...
+    
+    func _confirm() async -> (result: PaymentSheetResult, deferredIntentConfirmationType: STPAnalyticsClient.DeferredIntentConfirmationType?) {
+        let result: (PaymentSheetResult, STPAnalyticsClient.DeferredIntentConfirmationType?) = await {
+            
+            let authContext: STPAuthenticationContext? = {
+                switch configuration.formSheetAction {
+                case .confirm:
+                    if let formViewController {
+                        return formViewController
+                    }
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift:312]: nit: seems like we should check to make sure things are on screen before presenting?

```
                    if let formViewController,
                       formViewController.view.window != nil{
                        return formViewController
                    } else if let presentingViewController {
                        return STPAuthenticationContextWrapper(presentingViewController: presentingViewController)
                    }
                    return nil
```

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift [hunk 4]
```diff
@@ -293,6 +300,68 @@ extension EmbeddedPaymentElement: EmbeddedFormViewControllerDelegate {
...
+                                                    intent: intent,
+                                                    elementsSession: elementsSession,
+                                                    paymentOption: paymentOption,
+                                                    paymentHandler: paymentHandler,
+                                                    integrationShape: .embedded)
+        }()
+        
+        // Wait for the last update to finish and fail if didn't succeed. A failure means the view is out of sync with the intent and could e.g. not be showing a required mandate.
+        if let latestUpdateTask {
+            switch await latestUpdateTask.value {
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift:346]: I'm a bit confused on how this function works. It seems like you'd want to wait for latestUpdateTask to finish before attempting to confirm, but the way the code is written, it looks like `await PaymentSheet.confirm` could execute before checking `latestUpdateTask`?

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Link/PayWithLinkController.swift [hunk 5]
```diff
@@ -64,7 +64,7 @@ extension PayWithLinkController: PayWithLinkWebControllerDelegate {
             elementsSession: elementsSession,
             paymentOption: paymentOption,
             paymentHandler: paymentHandler,
-            isFlowController: false
+            integrationShape: .complete
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Link/PayWithLinkController.swift:67]: Maybe instead of passing in `integrationShape`, we just pass in the value for requiresInterstitialForCVC?

File: StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/EmbeddedPaymentElementTest.swift [hunk 6]
```diff
@@ -23,6 +23,19 @@ class EmbeddedPaymentElementTest: XCTestCase {
...
+    let paymentIntentConfigWithConfirmHandler = EmbeddedPaymentElement.IntentConfiguration(mode: .payment(amount: 1000, currency: "USD"), paymentMethodTypes: ["card", "cashapp"]) {paymentMethod, _, intentCreationCallback in
+        STPTestingAPIClient.shared.fetchPaymentIntent(types: ["card"],
+                                currency: "USD",
+                                paymentMethodID: paymentMethod.stripeId,
+                                confirm: true) { result in
+            switch result {
+            case .success(let clientSecret):
+                intentCreationCallback(.success(clientSecret))
+            case .failure(let error):
+                print(error)
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/EmbeddedPaymentElementTest.swift:35]: XCTFail("") ??

File: StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/EmbeddedPaymentElementTest.swift [hunk 7]
```diff
@@ -185,6 +212,68 @@ class EmbeddedPaymentElementTest: XCTestCase {
...
+        sut.delegate = self
+        sut.presentingViewController = UIViewController()
+        sut.view.autosizeHeight(width: 320)
+        
+        // Create test confirmParams with valid card details
+        let confirmParams = IntentConfirmParams(type: .stripe(.card))
+        confirmParams.paymentMethodParams.card = STPPaymentMethodCardParams()
+        confirmParams.paymentMethodParams.card?.number = "4242424242424242"
+        confirmParams.paymentMethodParams.card?.expMonth = NSNumber(value: 12)
+        confirmParams.paymentMethodParams.card?.expYear = NSNumber(value: 34)
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/EmbeddedPaymentElementTest.swift:228]: Maybe we add a common function like:
```
    func nextYear() -> Int {
        let year = Calendar.current.component(.year, from: Date())
        let yearPlus1 = year.advanced(by: 1)
        return yearPlus1 - 2000
    }
```

Comment (wooj-stripe) [reply StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/EmbeddedPaymentElementTest.swift:228]: We have some example of this already in `STPPaymentIntentFunctionalTest`

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift (hunk 1 above)
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift:27]: Renaming to `embeddedFormViewControllerDidCompleteConfirmation` and made a minor change to the doc comment.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift (hunk 3 above)
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift:312]: Added this, although I can't conceive a scenario where this would occur.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift (hunk 4 above)
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift:346]: `result` is a closure that gets [invoked](https://github.com/stripe/stripe-ios/pull/4242/files#diff-b66bb3ee17a6e8b6b2c784056b390b3ee60b34d788c38ba8aac66ef1ded1558dR349) if the last update has completed or just [returns](https://github.com/stripe/stripe-ios/pull/4242/files#diff-b66bb3ee17a6e8b6b2c784056b390b3ee60b34d788c38ba8aac66ef1ded1558dR362) result if there was no last update. We have a test for this [here](https://github.com/stripe/stripe-ios/pull/4242/files#diff-8908c6f71c82abec6b4f4989d9fa0ce5750cfa6e76702fb67d8c27ae86e9e1d7R175).

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Link/PayWithLinkController.swift (hunk 5 above)
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Link/PayWithLinkController.swift:67]: We did something similar to integration shape in PaymentSheetLoader, and I tend to prefer this over the flag as it's more extensible for possible future changes in confirmation depending on surface. The flag `isFlowController` is a good example of why it's limiting IMO.

File: Example/PaymentSheet Example/PaymentSheet Example/EmbeddedPlaygroundViewController.swift [hunk 8]
```diff
@@ -189,6 +190,24 @@ class EmbeddedPlaygroundViewController: UIViewController {
...
+            self.isLoading = true
+            let result = await embeddedPaymentElement.confirm()
+            self.isLoading = false
+            
+            switch result {
+            case .completed, .failed:
+                playgroundController?.lastPaymentResult = result
+                self.dismiss(animated: true)
+            case .canceled:
+                break
```
[the remark this answers is not in the document]
Comment (porter-stripe) [reply Example/PaymentSheet Example/PaymentSheet Example/EmbeddedPlaygroundViewController.swift:206]: This only happens when the user closes the sheet, I feel just dismissing the sheet is enough of an indication it was canceled and that's probably what I'd expect merchants to do in the real world too.

File: Example/PaymentSheet Example/PaymentSheetUITest/EmbeddedUITest.swift [hunk 9]
```diff
@@ -101,6 +101,13 @@ class EmbeddedUITests: PaymentSheetUITestCase {
         XCTAssertTrue(app.buttons["Reload"].waitForExistence(timeout: 10))
         // ... Klarna should still be selected
         XCTAssertEqual(app.staticTexts["Payment method"].label, "Klarna")
+        
+        // Confirm the Klarna payment
+        XCTAssertTrue(app.buttons["Checkout"].isEnabled)
+        app.buttons["Checkout"].waitForExistenceAndTap()
```
Comment (wooj-stripe) [inline Example/PaymentSheet Example/PaymentSheetUITest/EmbeddedUITest.swift:107]: Check existence before checking isEnabled?

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift (hunk 4 above)
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift:307]: You're understanding is right, mine wasn't. Fixed this.

File: StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/EmbeddedPaymentElementTest.swift [hunk 10]
```diff
@@ -138,14 +151,28 @@ class EmbeddedPaymentElementTest: XCTestCase {
...
+        let sut = try await EmbeddedPaymentElement.create(intentConfiguration: paymentIntentConfigWithConfirmHandler, configuration: configuration)
+        sut.presentingViewController = UIViewController()
+        sut.view.autosizeHeight(width: 320)
+        
+        // Create test confirmParams with valid card details
+        let confirmParams = IntentConfirmParams(type: .stripe(.card))
+        confirmParams.paymentMethodParams.card = STPPaymentMethodCardParams()
+        confirmParams.paymentMethodParams.card?.number = "4242424242424242"
+        confirmParams.paymentMethodParams.card?.expMonth = NSNumber(value: 12)
+        confirmParams.paymentMethodParams.card?.expYear = NSNumber(value: 34)
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/EmbeddedPaymentElementTest.swift:163]: extend +5 years? (doesnt need to be in this pr)

**Review by wooj-stripe — APPROVED** [review]
